File: frontend/pali-studio/src/types/literature.ts
/**
 * Literature and Segment TypeScript types
 */

export interface Literature {
  id: string;
  name: string;
  pali_name: string;
  pitaka: string;
  nikaya: string | null;
  status: 'translated' | 'parsed';
  total_segments: number;
  translated_segments: number;
  source_pdf: string | null;
  hierarchy_labels: Record<string, string> | null;
  display_metadata: LiteratureDisplayMetadata | null;
  created_at: string;
  updated_at: string;
}

export interface LiteratureDisplayMetadata {
  ko_translit?: string;
  abbr?: string;
  aliases?: string[];
  description?: string;
}

export interface TranslationSentence {
  original_pali: string;
  grammatical_analysis: string;
  literal_translation: string;
  free_translation: string;
  explanation: string;
}

export interface Translation {
  sentences: TranslationSentence[];
  summary?: string;
}

export interface Segment {
  id: number;
  literature_id: string;
  vagga_id: number | null;
  vagga_name: string | null;
  sutta_id: number | null;
  sutta_name: string | null;
  page_number: number | null;
  paragraph_id: number;
  original_text: string;
  translation: Translation | null;
  is_translated: boolean;
  created_at: string;
  updated_at: string;
}

export interface PaginationMeta {
  offset: number;
  limit: number;
  total: number;
  has_more: boolean;
}

export interface LiteratureListResponse {
  literatures: Literature[];
  pitaka_structure: PitakaStructure;
  total_count: number;
}

export interface SegmentListResponse {
  segments: Segment[];
  pagination: PaginationMeta;
}

export interface PitakaStructure {
  sutta: Record<string, string[]>;
  vinaya: string[];
  abhidhamma: string[];
}

export interface VaggaInfo {
  vagga_id: number;
  vagga_name: string;
  sutta_count: number;
}

File: backend/pali/schemas/literature.py
"""Pydantic schemas for literature-related endpoints."""
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field

from .common import PaginationMeta


class LiteratureBase(BaseModel):
    """Base schema for literature data."""
    name: str
    pali_name: str
    pitaka: str
    nikaya: Optional[str] = None
    source_pdf: Optional[str] = None
    hierarchy_labels: Optional[Dict[str, str]] = None
    display_metadata: Optional[Dict[str, Any]] = None


class LiteratureCreate(LiteratureBase):
    """Schema for creating a new literature."""
    id: str


class LiteratureResponse(LiteratureBase):
    """Schema for literature response."""
    id: str
    status: str
    total_segments: int
    translated_segments: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class LiteratureListResponse(BaseModel):
    """Schema for literature list response."""
    literatures: List[LiteratureResponse]
    pitaka_structure: Dict[str, Any]
    total_count: int


class SegmentBase(BaseModel):
    """Base schema for segment data."""
    vagga_id: Optional[int] = None
    vagga_name: Optional[str] = None
    sutta_id: Optional[int] = None
    sutta_name: Optional[str] = None
    page_number: Optional[int] = None
    paragraph_id: int
    original_text: str


class SegmentResponse(SegmentBase):
    """Schema for segment response."""
    id: int
    literature_id: str
    translation: Optional[Dict[str, Any]] = None
    is_translated: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SegmentListResponse(BaseModel):
    """Schema for segment list response with pagination."""
    segments: List[SegmentResponse]
    pagination: PaginationMeta


class VaggaInfo(BaseModel):
    """Information about a vagga (section)."""
    vagga_id: int
    vagga_name: str
    sutta_count: int


class PitakaStructure(BaseModel):
    """Tripitaka structure for navigation."""
    sutta: Dict[str, List[str]]  # nikaya -> literature_ids
    vinaya: List[str]
    abhidhamma: List[str]
